api/src/data_migration/transformation/subtask/transform_competition.py

```diff
@@ -0,0 +1,236 @@
+
+import src.data_migration.transformation.transform_constants as transform_constants
+import src.data_migration.transformation.transform_util as transform_util
+from src.constants.lookup_constants import CompetitionOpenToApplicant, FormFamily
+from src.data_migration.transformation.subtask.abstract_transform_subtask import (
+    AbstractTransformSubTask,
+)
+from src.db.models.competition_models import Competition
+from src.db.models.opportunity_models import Opportunity
+from src.db.models.staging.competition import Tcompetition
+from src.db.models.staging.opportunity import TopportunityCfda
+
+logger = logging.getLogger(__name__)
+
+
+class TransformCompetition(AbstractTransformSubTask):
+    def transform_records(self) -> None:
+        logger.info("Processing competitions")
+
+        # Fetch all competitions that need to be processed
+        # Alongside that, grab the existing competition record
+        competitions: list[tuple[Tcompetition, Competition | None]] = self.fetch(
+            Tcompetition,
+            Competition,
+            [Tcompetition.comp_id == Competition.legacy_competition_id],
+        )
+
+        for source_competition, target_competition in competitions:
+            try:
+                self.process_competition(source_competition, target_competition)
+            except ValueError:
+                self.increment(
+                    transform_constants.Metrics.TOTAL_ERROR_COUNT,
+                    prefix=transform_constants.COMPETITION,
+                )
+                logger.exception(
+                    "Failed to process competition",
+                    extra={"competition_id": source_competition.comp_id},
+                )
+
+    def process_competition(
+        self, source_competition: Tcompetition, target_competition: Competition | None
+    ) -> None:
+        self.increment(
+            transform_constants.Metrics.TOTAL_RECORDS_PROCESSED,
+            prefix=transform_constants.COMPETITION,
+        )
+
+        extra = {"competition_id": source_competition.comp_id}
+        logger.info("Processing competition", extra=extra)
+
+        if source_competition.is_deleted:
+            self._handle_delete(
+                source=source_competition,
+                target=target_competition,
+                record_type=transform_constants.COMPETITION,
+                extra=extra,
+            )
+            # Update transformed_at timestamp to mark it as processed
+            source_competition.transformed_at = self.transform_time
+            return
+
+        # Find the related opportunity and opportunity_assistance_listing
+        opportunity_id = None
+        opportunity_assistance_listing_id = None
+
+        # One major thing to call out is how a competition connects to an opportunity in the old system.
+        # These always needed to go via opportunity_cfda (now opportunity assistance listing).
+        if source_competition.opp_cfda_id is not None:
```

Comment:
Could we invert this and raise an error if it happens? While the column is nullable, at least in prod there aren't any cases of it actually being null, so it's definitely an edge case where we should be made aware of it.


```suggestion
        # One major thing to call out is how a competition connects to an opportunity in the old system.
        # These always needed to go via opportunity_cfda (now opportunity assistance listing).
        if source_competition.opp_cfda_id is None:
             raise ValueError("Some sort of message")
        ...
```